### ISSUE TITLE
Design: Bulk Get API

### ISSUE THREAD

**Issue by yaron2** [dapr/dapr#1728]
This is the design proposal for the Bulk Get API.

Two tracking and semi-related issues can be found here:

https://github.com/dapr/dapr/issues/1338
https://github.com/dapr/dapr/issues/1339

# Scope

The scope of this feature is a subset of #1339. As discussed in the various issues and on maintainer meeting, we will first begin with the requested feature for being able to give Dapr a list of keys and get back a corresponding list of values.

This is good for high performance scenarios and for allowing the client to create basic filtering and querying of data on the application side.

An example scenario to validate the business case is the all time favourite shopping cart:

A stateless service saves 500 items, where each item's key is the cart id.
Later, A business process in the app needs to be able to get back all 500 items because they were all either corrupted, faulty or labeled in any way that justifies pulling them.

Without a Bulk Get API, the user app would need to make 500 round trips to the Dapr sidecar.
With a Bulk Get API, the user app would need to make a single round trip to the Dapr sidecar.

The performance penalty associated with the former way of getting the state increases with the number of items, making it O(n) in terms of the number of hops from the app to the Dapr sidecar.

## Proposal

Dapr will expose two new endpoints, one for HTTP and one for gRPC.

### HTTP endpoint

```
v1.0/state/my-store/bulk POST
```

The `POST` verb calls for a more detailed and thorough explanation here.
Naturally, it would seem that the `GET` verb is more appropriate.
However on close inspection of HTTP semantics, we deduce the following:

In order to pass a list of keys in a `GET` query, one would need to use query parameters. While there is no official limit of characters with regards to the HTTP spec, many web server have defaults like max 1024/2048 characters.

While it is possible for Dapr to configure `fasthttp` to ignore those limits, it feels wrong to have users construct a URL that looks like `http://localhost:3500/v1.0/state/?keys=1,2,3,4,5,6,7,8,9,0,1,2,3,4,5,6,7,8,9,0,1,2,3,4,5,6,7,8,9,0,1,2,3,4,5,6,7,8,9,0,1,2,3,4,5,6,7,8,9,0,1,2,3,4,5,6,7,8,9,0,1,2,3,4,5,6,7,8,9,0,1,2,3,4,5,6,7,8,9,0,1,2,3,4,5,6,7,8,9,0,1,2,3,4,5,6,7,8,9,0,1,2,3,4,5,6,7,8,9,0,1,2,3,4,5,6,7,8,9,0,1,2,3,4,5,6,7,8,9,0,1,2,3,4,5,6,7,8,9,0,1,2,3,4,5,6,7,8,9,0,1,2,3,4,5,6,7,8,9,0.....N`.

As such, it is recommended to use a `POST` body with a well defined schema to represent the keys to be fetched, and more closely aligns with the corresponding proto service definition detailed below.

#### Request JSON schema example

The API will receive an array of strings, each item in the list represents a key.

```json
["1","2","3","4","5","6","7","8","9","0","1","2",3","4","5","6","7","8","9","0","1","2","3",4","5","6","7","8","9","0","1","2","3","4","5","6","7","8","9","0","1","2","3","4"]
```

#### Response JSON schema example

The response will returns the following structure:

```json
[
  {
     "key": 1,
     "value": "hello there"
  },
 {
     "key": 2,
     "value": "general kenobi"
  }
]
```

### gRPC method

The proto service definition for `v1.runtime` will be updated with a new service operation:

```proto
rpc GetBulkState(GetBulkStateRequest) returns (GetBulkStateResponse) {}

message GetBulkStateRequest {
  repeated string keys = 1;
}

message GetBulkStateResponse {
  repeated BulkStateItem keys = 1;
}

message BulkStateItem {
  string key = 1;
  bytes data = 2;
}
```

### Relationship with `components-contrib`

The Dapr sidecar will receive the keys in a single command, and then proceed to create concurrent go-routines to fetch the items from the selected state store via the regular `Get` method.

This implementation requires no change to the state store implementations in components-contrib.
This enables this feature to apply to every Dapr state store, without forcing specific implementations to support Bulk Get operations.

In the future, we can look into adding Bulk Get methods on the `components-contrib` interface and allow a state store to specify whether or not it supports that operation. This will allow Dapr to use native bulk capabilities of the store where possible and fallback to single fetch operations where needed.

**Comment by pruthvidhodda** [comment]
Should we consider having ETags in responses?

**Comment by Haishi2016** [comment]
1) Should we factor certain query capabilities, such as wildcard/pattern matching, into this?
2) Should we consider pagination of larger datasets with something like continuation tokens?

**Comment by rynowak** [comment]
## High level

> Without a Bulk Get API, the user app would need to make 500 round trips to the Dapr sidecar.
> With a Bulk Get API, the user app would need to make a single round trip to the Dapr sidecar.

...

> The Dapr sidecar will receive the keys in a single command, and then proceed to create concurrent go-routines to fetch the items from the selected state store via the regular Get method.

Are we sure that this really solves the problem? It seems like the expensive parts of the operation are still O(N) without the feature being implemented end-to-end to the state store. 

- Serialization is O(N) (can't be avoided)
- Network hop to the sidecar is O(1)
- Network hop from sidecar -> backing state store is O(N) without end-to-end impl
- Backing state store has to respond to O(N) queries without end-to-end impl

The things that get better are: surfacing a first-class API in the SDKs, and optimizing the smaller of our two network hops.

---

This also seems like it encourages patterns that are non-consistent by default, while making it seem like it might be a consistent operation.

**Case 1:** You query the user's cart to get the list of item IDs, then issue another query to get all of the items (by id).

This isn't fundamentally the same query as something like (pseudocode):

```
get where id is in(cart.items)
```

because it's not a single consistent operation. 

You can observe changes to the items that took place in between the time you retrieved the Ids and the time you retrieved the items. I have trouble thinking of a specific schema/scenario where this would get you into trouble, but I wanted to point out that issuing multiple reads to a state store is not semantically equivalent to issuing a single query.

**Case 2:** (same scenario as above)

You can observe changes to the items that took place while the "query" you issued with the SDK was running, because the query to get all of the items is *actually N queries*. If the items have dependencies or relationships, then you can end up with an inconsistent set of items, because they were retrieved atomically.

It could probably be easy to say *"don't do that"* for these cases, but it seems worth pointing out that both the pattern *get keys and then get items*, and the implementation *get each item individually* have different semantics than real query support. The second case could be addressed by adding end-to-end support to the state store, the first case seems inherent. 

## Minor stuff

We should specify what happens when a key is not found. Null? Error?

---

For the JSON protocol - it seems wise to turn the request and response into objects instead of arrays.  You can't add a property to an array (in JSON) so you can't add functionality without a breaking change to the protocol. The protobuf definitions don't have the same issue.

One of the other things you *could* do (after making these objects) is allow state store implementations to extend this with their own support via their own property definitions. This would be per-provider and implementation specific.

**Comment by yaron2** [comment]
> Should we consider having ETags in responses?

Yes, the responses should have etags, I'll update the proposal.

**Comment by yaron2** [comment]
> 1. Should we factor certain query capabilities, such as wildcard/pattern matching, into this?

So this is a good suggestion that'd more closely discussed here: #1339, but it out of scope for this piece of work as this is focused on a feature that can be implemented across all state stores. 

> 2. Should we consider pagination of larger datasets with something like continuation tokens?

Yes, probably. CosmosDB for example will most likely return continuation tokens on most large multi-key fetches.

**Comment by yaron2** [comment]
> Are we sure that this really solves the problem? It seems like the expensive parts of the operation are still O(N) without the feature being implemented end-to-end to the state store.

This talks specifically about the app to sidecar calls, which will be O(1). I updated the proposal to make that clear.

> This also seems like it encourages patterns that are non-consistent by default, while making it seem like it might be a consistent operation.

Bulk get is for eventual consistent scenarios, this assumption was validated with the parties requesting this specific feature.

**Comment by Haishi2016** [comment]
Looking at #1339, I feel this feature should be folded into that item, or at least thinking of a path to merge the two. The oData layer shall cross state stores as well, I presume. 

As for consistency, if we retrieve id list first then use the list to retrieve data, we will have the consistency problem as @rynowak laid out. However this is not the case. Presumably the ids are fed to the underlying data store in one query, which generates the snapshot to be returned using data store's default isolation mechanism.

**Comment by yaron2** [comment]
> As for consistency, if we retrieve id list first then use the list to retrieve data, we will have the consistency problem as @rynowak laid out. However this is not the case. Presumably the ids are fed to the underlying data store in one query, which generates the snapshot to be returned using data store's default isolation mechanism.

That's correct @Haishi2016, also 'get items` and 'get values' are not distinct operations. the query is always `get values for ids`.

> Looking at #1339, I feel this feature should be folded into that item, or at least thinking of a path to merge the two. The oData layer shall cross state stores as well, I presume.

It is tracked as part of #1339, but we want break the work down into smaller pieces. As there are multiple asks for getting a list of items which do not necessitate full OData like query capabilities.

**Comment by rynowak** [comment]
> This talks specifically about the app to sidecar calls, which will be O(1). I updated the proposal to make that clear.

I think the proposal was already pretty clear on that point, but I'm questioning whether it's worth building the feature without the plumbing to do it E2E. This is the classic N+1 query problem baked into Dapr behind a first-class API 😁 

Do we really think that eliminating the localhost network hop to the sidecar is enough to encourage this pattern?

**Comment by rynowak** [comment]
> The item is missing from the response.

This seems like it should work pretty well 👍

**Comment by yaron2** [comment]
> Do we really think that eliminating the localhost network hop to the sidecar is enough to encourage this pattern?

Well, I think we can agree that one network hop as opposed to 1000 is more performant, so in that regards, yes I am quite certain this will reduce latencies. The issue here is not about eliminating a problem, its about enabling users to do the same query more efficiently and with less boilerplate code.

**Comment by rynowak** [comment]
> Well, I think we can agree that one network hop as opposed to 1000 is more performant, so in that regards, yes I am quite certain this will reduce latencies. The issue here is not about eliminating a problem, its about enabling users to do the same query more efficiently and with less boilerplate code.

I think we agree on all of the technical aspects of this. If you as a user are determined to do a bulk retrieval already then having this in the SDK makes your life better.

With respect, I think I have a different perspective on removing the boilerplate code. To me, putting a first class feature into an SDK is an advertisement to "do it like this". Users will see that API and conclude that they *should* use it - they will be happy that someone thought about their use case and included it - they will design their schema and usage *based on* the existence of the APIs. 

We should take responsibility for the consequences of including a first class feature that removes an O(N) from user code and hides it behind an abstraction. If users encounter bad performance and feel like they were coaxed into a snafu, then that seems totally fair 😈 

I'm not going to argue this point too fiercely since I'm coming in without a lot of context - I would feel a lot better knowing that making "bulk get" end-to-end is something we'd consider. 

---

> In addition, I wouldn't call this a pattern, as this mimics standard behaviors you see on different databases, being able to get a list of items based on a list of IDs.

I'll agree that you *could* write this code on any DB today, but the difference is that you don't have to 😆

**Comment by yaron2** [comment]
> I'm not going to argue this point too fiercely since I'm coming in without a lot of context - I would feel a lot better knowing that making "bulk get" end-to-end is something we'd consider.

Sure, as the proposal points out, extending the components-contrib interface for bulk-get can happen going forward and using it for the state stores that support it.

> I'll agree that you could write this code on any DB today, but the difference is that you don't have to 😆

Well, but you don't have to do it with Dapr either.. just like coding with a DB SDK and a Dapr SDK, you have Get and GetBulk operations on both :)

**Comment by yaron2** [comment]
Closed via #1888
